docs/designdoc/support_volume_expansion.md

```diff
@@ -0,0 +1,89 @@
+# Support Volume Expansion Through MySQLCluster
+
+## Context
+
+Support PVC expansion in MOCO.
+
+MOCO is using StatefulSet to build a MySQL cluster.
+Currently, the StatefulSet PVC template is not editable.
+
+Users can expand PVCs by performing manual operations:
+
+> 1. Resize all PVCs manually
+> 2. Edit MySQLCluster and change the requested volume size
+> 3. Delete StatefulSet w/o deleting Pods
+>     - `kubectl delete sts moco-xxx --cascade=false`
+>
+> ref: https://github.com/cybozu-go/moco/issues/265
+
+The moco-controller supports this operation so that the user does not have to do it.
+
+> :note: It may be possible to edit the StatefulSet directly in the future.
+> 
+> ref: https://github.com/kubernetes/enhancements/issues/661
+
+## Goals
+
+* Automatically extend the volume of PVCs used by MySQLCluster
+* No breaking changes
+
+## Non-goals
+
+* No enhancement will be added to the `moco.cybozu.com/v1beta1` API.
+
+## ActualDesign
+
+moco-controller automatically resizes the PVC when the size of the MySQLCluster volume claim is extended.
+
+The steps are as follows:
+
+1. Checks if `allowVolumeExpansion` is `true` for the StorageClass specified in the PVC
+2. Resize all PVCs
+3. Delete only the StatefulSet without deleting the Pods
+4. StatefulSet will be recreated with the new PVC size
+5. (Optional) Restart the StatefulSet
+
+The functions that must be performed to achieve this functionality are steps 1 through 4.
+Step 5 is not necessary if the volume plugin supports [online file system expansion](https://kubernetes.io/blog/2018/07/12/resizing-persistent-volumes-using-kubernetes/#online-file-system-expansion).
+
+### Validation
+
+Currently, MOCO does not validate against `.spec.volumeClaimTemplates`.
+This time add the following validate:
+
+* Make volumeClaimTemplates invariant except for storage size
+* Allow only incremental changes in storage size
+  * If storage expansion is involved, the storage class `allowVolumeExpansion` is set to `true`.
+  * Refer to the storage size in the StatefulSet to compare with the actual size.
+
+### Resize PVCs
+
+When moco-controller resizes a PVC, there may be a discrepancy between the PVC defined in the MySQLCluster and the actual PVC size.
+For example, if you are using [github.com/topolvm/pvc-autoresizer](https://github.com/topolvm/pvc-autoresizer).
+In this case, moco-controller will only update if the actual PVC size is smaller than the PVC size after the change.
+
+Failure to update will result in a discrepancy between the PVC defined in the MySQLCluster and the actual PVC.
+Add the condition `VolumeResized` to the MySQLCluster condition to signal this condition.
+
+```yaml
+kind: MySQLCluster
+status:
+  conditions:
+    - lastTransitionTime: "2022-05-01T16:29:24Z"
+      status: "False"
+      type: VolumeResized
+      message: "Validation failed: ..."
+```
+
+Set the `VolumeResized` condition's status to `True` if the PVC expansion was successful.
```

Comment:
That fine with me.
Could you also export this status as a metric?

Reply:
I see, I came up with two options, which would you prefer?

1. Count metric, `moco_pvc_expansion_failed`
    * Count is incremented if PVC resizing fails
2. Gauge metric, `moco_pvc_expansion_status{states="True/False"}`
    * Aggregate the number of objects from the list of MySQLCluster's that currently exist with that status

Reply:
Could you export this metrics for each MySQLCluster?
And also use the prefix of `moco_cluster_` as with other metrics.
https://github.com/cybozu-go/moco/blob/main/docs/metrics.md#mysql-clusters

How about use the following format?
Count metircs: `moco_cluster_volume_resized_total{status="True/False", name="mycluster", namespace="mynamesapce"}`

Reply:
@masa213f 
As I told you elsewhere, moco-controller should re-create StatefulSet not only when the storage size is changed, but also when metadata (labels / annotations) is changed. Considering that, the metric name containing `volume_resized` is not appropriate.

Reply:
@d-kuro Have you heard about that?

Reply:
I updated the docs to reflect the comments 🙏 :
https://github.com/cybozu-go/moco/pull/412/commits/e63a1d56f2c81684fc0c7492e43058c1538c4735